pkg/provider/azure_vmssflex_cache.go

```diff
@@ -0,0 +1,309 @@
+			if rerr != nil {
+				if rerr.IsNotFound() {
+					klog.Warningf("Skip caching vmss for resource group %s due to error: %v", resourceGroup, rerr.Error())
+					continue
+				}
+				klog.Errorf("VirtualMachineScaleSetsClient.List failed: %v", rerr)
+				return nil, rerr.Error()
+			}
+
+			for i := range allScaleSets {
+				scaleSet := allScaleSets[i]
+				if scaleSet.ID == nil || *scaleSet.ID == "" {
+					klog.Warning("failed to get the ID of VMSS Flex")
+					continue
+				}
+
+				if scaleSet.OrchestrationMode == compute.OrchestrationModeUniform {
+					klog.V(2).Infof("Skip Uniform VMSS: (%s)", *scaleSet.ID)
+					continue
+				}
+
+				localCache.Store(*scaleSet.ID, &scaleSet)
+			}
+		}
+
+		return localCache, nil
+	}
+
+	if fs.Config.VmssFlexCacheTTLDefaultInSeconds == 0 {
+		fs.Config.VmssFlexCacheTTLDefaultInSeconds = consts.VmssFlexCacheTTLDefaultInSeconds
+	}
+	return azcache.NewTimedcache(time.Duration(fs.Config.VmssFlexCacheTTLDefaultInSeconds)*time.Second, getter)
+}
+
+func (fs *FlexScaleSet) newVmssFlexVMCache() (*azcache.TimedCache, error) {
+	getter := func(key string) (interface{}, error) {
+		localCache := &sync.Map{}
+
+		ctx, cancel := getContextWithCancel()
+		defer cancel()
+
+		vms, rerr := fs.VirtualMachinesClient.ListVmssFlexVMsWithoutInstanceView(ctx, key)
+		if rerr != nil {
+			klog.Errorf("ListVmssFlexVMsWithoutInstanceView failed: %v", rerr)
+			return nil, rerr.Error()
+		}
+
+		for i := range vms {
+			vm := vms[i]
+			if vm.Name != nil {
+				localCache.Store(*vm.Name, &vm)
+				fs.vmssFlexVMNameToVmssID.Store(*vm.Name, key)
+			}
+		}
+
+		return localCache, nil
+	}
+
+	if fs.Config.VmssFlexVMCacheTTLInSeconds == 0 {
+		fs.Config.VmssFlexVMCacheTTLInSeconds = consts.VmssFlexVMCacheTTLInSeconds
+	}
+	return azcache.NewTimedcache(time.Duration(fs.Config.VmssFlexVMCacheTTLInSeconds)*time.Second, getter)
+}
+
+func (fs *FlexScaleSet) newVmssFlexVMStatusCache() (*azcache.TimedCache, error) {
+	getter := func(key string) (interface{}, error) {
+		localCache := &sync.Map{}
+
+		ctx, cancel := getContextWithCancel()
+		defer cancel()
+
+		vms, rerr := fs.VirtualMachinesClient.ListVmssFlexVMsWithOnlyInstanceView(ctx, key)
+		if rerr != nil {
+			klog.Errorf("ListVmssFlexVMsWithOnlyInstanceView failed: %v", rerr)
+			return nil, rerr.Error()
+		}
+
+		for i := range vms {
+			vm := vms[i]
+			if vm.Name != nil {
+				localCache.Store(*vm.Name, &vm)
+				fs.vmssFlexVMNameToVmssID.Store(*vm.Name, key)
+			}
+		}
+
+		return localCache, nil
+	}
+
+	if fs.Config.VmssFlexVMStatusCacheTTLInSeconds == 0 {
+		fs.Config.VmssFlexVMStatusCacheTTLInSeconds = consts.VmssFlexVMStatusCacheTTLInSeconds
+	}
+	return azcache.NewTimedcache(time.Duration(fs.Config.VmssFlexVMStatusCacheTTLInSeconds)*time.Second, getter)
+}
+
+func (fs *FlexScaleSet) getNodeVmssFlexID(nodeName string) (string, error) {
+	fs.lockMap.LockEntry("getNodeVmssFlexID")
```

Comment:
move L144-L145 to L148

Reply:
`getNodeVmssFlexID` should be defined as a const key

Reply:
A new key is assed to the consts.

Reply:
I don't think it should be moved. Let's assume if 2 nodes are added at the same time, if moving the lock into the if !cached block, then the cache will be refreshed twice. Instead, keeping the lock outside of the if logic can avoid such issue. Let me know if I missed anything.